### PR TITLE
Add context menu on layer list, introduce `QtActionContextMenu`.

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
This is a work-in-progress PR that will add a context menu to the layer list... plenty to discuss here on the API and organization (and the code itself is still a mess...) but here's a rough idea of what we can do with it.  Note that the options in the dropdown menu can change based on what is selected:

https://user-images.githubusercontent.com/1609449/114798738-091b3300-9d64-11eb-926e-da4b9a37a9df.mov

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by jni** [comment]
:heart_eyes: hot damn, I love this @tlambert03!

**Comment by tlambert03** [comment]
some updates here that I think might be a nice pattern... curious to get some feedback:

1. items in the context menu can now be added in a rather declarative manner using `LAYER_ACTIONS` [here](https://github.com/napari/napari/pull/2556/files#diff-568b9a696950cc6898d5a4ca3dc48db51f9009243ab8d8cc4038cfe2721ac28cR64-R120)... each value in the dict has mostly the same parameters as `action_manager.register_action`, with the addition of a `when` key that acts as the "context" for when the item should be enabled.  The `action` key is a function that takes an instance of `LayerList` and returns a value (doesn't have to return a boolean though).
2. The `when` field is an expression (string) that is allowed to use any of the keys named in the `CONTEXT_KEYS` dict [here](https://github.com/napari/napari/pull/2556/files#diff-acfb43cfe7047ea39e1d0f252aa2225d6a7f3f2f1fd59e11efb7851c219427d1R336-R352).  for example: `'selection_count > 1 and not all_layers_linked'`
3. Each of the values in `CONTEXT_KEYS` is a function that takes the current `layerlist.selection` and returns a value.
4. the `LayerList` gains a new method (`selection_context`) [here](https://github.com/napari/napari/pull/2556/files#diff-acfb43cfe7047ea39e1d0f252aa2225d6a7f3f2f1fd59e11efb7851c219427d1R332-R333) that turns the `CONTEXT_KEYS` dict into a concrete context map.
4.  The actual context menu itself becomes a [rather small class](https://github.com/napari/napari/pull/2556/files#diff-9befe15c3860c4a5a3aae45e3c54dfaac74a2aeb10b5d1c7c0849e906b7527d7R8-R25) that can turn one of these actions dicts into a `QMenu`.  
5. The layer delegate [creates the context menu](https://github.com/napari/napari/pull/2556/files#diff-eeac0e32eb670a45245c01e4d5ad46e711a00bed172205dc195818bedd177e77R178-R185), and whenever it is right clicked, it will use the [`_update` method](https://github.com/napari/napari/pull/2556/files#diff-9befe15c3860c4a5a3aae45e3c54dfaac74a2aeb10b5d1c7c0849e906b7527d7R18-R25) on the action QMenu, which performs the actual job of evaluating all of the `when` expressions to determine if each item in the menu should be enabled or not.
6. finally, if one of the items is clicked,  the corresponding `action` is called with the current layer_list as a argument

**Comment by tlambert03** [comment]
I've added some tests and a lot more documentation.  this is probably ready for some eyes @jni, @sofroniewn... but may well need some more modifications and/or documentation before it's ready.

File: napari/layers/utils/_link_layers.py
Comment (tlambert03) [inline napari/layers/utils/_link_layers.py:29]: this new `get_linked_layers` function that takes some layers and tells you what they are linked to.  it required changing the way the link-keys are stored from layer `ids` to layer `weakrefs` ... hence some of the other changes in this module.

File: napari/layers/base/base.py
Comment (tlambert03) [inline napari/layers/base/base.py:1142]: `Layer.create` is a classmethod that takes the arguments in a LayerDataTuple and makes a new layer.  Seems like something we could generally use, and is useful here for the duplicate layer command.  we really ought to be able to use it to replace the `ViewerModel._add_layer_from_data` method... but the fact that `add_image` takes `channel_axis` screws up that otherwise clean "create one layer then add it" pattern.

**Comment by sofroniewn** [comment]
Just tried this out and got the following error. Will review code now too

```python
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
~/GitHub/napari/napari/_qt/containers/_layer_delegate.py in editorEvent(self=, event=, model=, option=, index=)
    151         ):
    152             self.show_context_menu(
--> 153                 index, model, event.globalPos(), option.widget
        index = 
        model = 
        event.globalPos = 
        option.widget = None
    154             )
    155 

~/GitHub/napari/napari/_qt/containers/_layer_delegate.py in show_context_menu(self=, index=, model=, pos=PyQt5.QtCore.QPoint(188, 454), parent=)
    180 
    181         layer_list: LayerList = model.sourceModel()._root
--> 182         self._context_menu._update(layer_list.selection_context())
        self._context_menu._update = undefined
        layer_list.selection_context = , , , , , ]>
    183         action = self._context_menu.exec_(pos)
    184         if action is not None:

AttributeError: 'QtActionContextMenu' object has no attribute '_update'
```

File: napari/layers/base/base.py
Comment (sofroniewn) [reply napari/layers/base/base.py:1142]: > but the fact that add_image takes channel_axis screws up that otherwise clean "create one layer then add it" pattern.

It is a little funny. We could deprecate and make a two line API, first add, then split and have GUI users be able to use context menu. Would you be supportive @jni? It would simplify a lot of code. It's also a pattern we've decided to use no where else in the code base.

File: napari/components/layerlist.py
Comment (sofroniewn) [inline napari/components/layerlist.py]: Love this, but can we make it private, `_CONTEXT_KEYS`. I suspect we might refine some of these names, but don't want to get hung up on them now. If you think important to be public we can talk through more of them. The most confusing one for me is `active_shape`, not quite clear what that one is just by reading it

File: napari/layers/_layer_actions.py
Comment (sofroniewn) [inline napari/layers/_layer_actions.py:118]: Can you add `'show_when': 'True'` to all the places where it is missing. I find the extra line removes all ambiguity - it's much easier for me to read that and know it's always shown then have to remember if nothing specified then it's always shown. I think I might require all 4 params, `'description', 'action', 'enable_when', 'show_when'` and throw an error if one is missing. Just keeps things simple

File: napari/utils/events/event.py
Comment (sofroniewn) [inline napari/utils/events/event.py:261]: No problem with these additions, but curious why they are needed - was there a bug before? Maybe you can add a comment in the code

File: napari/layers/_layer_actions.py
Comment (sofroniewn) [inline napari/layers/_layer_actions.py]: Do we need translations on these strings? not sure if that's so easy here?

**Review by sofroniewn — COMMENTED** [review]
Overall this is looking great to me, really like how it's shaping up. It'll provide a lot of value to a lot of folks!

I'd love for @Carreau, @ppwadhwa, and @goanpeca to review this too before it goes in, ideally we'd have as much in common for our "action" concept as possible - for example if I wanted to bind a shortcut to one of these actions hopefully it wouldn't be too hard to eventually make it appear in the UI that @ppwadhwa is building in #2864 too (maybe we'd think of these as actions (keybindings) bound to the layerlist selection object, as in a sense that's sort of what they are)

File: napari/layers/_layer_actions.py
Comment (tlambert03) [reply napari/layers/_layer_actions.py:118]: I can, but it will incur an extra (unnecessary) `eval` for each one of them.  Probably negligible overhead, but that's why I left it out.  I'll test it

File: napari/layers/_layer_actions.py
Comment (tlambert03) [reply napari/layers/_layer_actions.py]: Ugh... yeah probably

**Comment by tlambert03** [comment]
> AttributeError: 'QtActionContextMenu' object has no attribute '_update'


Darn, sorry.  Changed that name to `update_from_context` and forgot to update there

File: napari/components/layerlist.py
Comment (tlambert03) [reply napari/components/layerlist.py]: Sure, private is fine.

Will change to `active_layer_shape`

File: napari/layers/_layer_actions.py
Comment (tlambert03) [reply napari/layers/_layer_actions.py:118]: made them all mandatory.  (No runtime error if one is missing, but added a test)

File: napari/utils/events/event.py
Comment (tlambert03) [reply napari/utils/events/event.py:261]: I needed them to avoid an emitter loop in [`selectionChanged`](https://github.com/napari/napari/pull/2556/files#diff-1d29b7b2ae740db911dd210ab47b659628e811dd428b5aa51ea3517fb23cbe1bR77).  this variable is actually present in the [original vispy events.py](https://github.com/vispy/vispy/blob/184d854c7542f4c6df21bb4c7db3dd5b340dd900/vispy/util/event.py#L431) ... it must have gotten removed from our version, but I think it's actually useful

**Comment by tlambert03** [comment]
> ideally we'd have as much in common for our "action" concept as possible

yep, that was the general goal with the design here.  Adding keybinding could be as easy as adding a `'shortcut'` to each action in `LAYER_ACTIONS`.  I agree, they're essentially actions bound to the `layerlist.selection`.

made all the requested changed @sofroniewn... will wait for further reviews

File: napari/utils/events/event.py
Comment (sofroniewn) [reply napari/utils/events/event.py:261]: hmm funny, not sure what happened there!

**Comment by sofroniewn** [comment]
This is so much fun to play with. I can get a reproducible error with the following actions
- run python examples/layers.py
- duplication the blobs layer
- select the blobs layer and the random layer, and click merge

If I don't duplicate the blobs layer first then the merge works fine

The full stack trace is very long, here's the last part

```python-traceback
~/GitHub/napari/napari/utils/events/event.py in _invoke_callback(self=<napari.utils.events.event.EmitterGroup object>, cb=<bound method EventedList._reemit_child_event of...dec50>, <Image layer 'blobs' at 0x7fdabf9dec90>]>, event=<Event blocked=False handled=False native=None source=None sources=[] type='translate'>)
    594     def _invoke_callback(self, cb: Callback, event: Event):
    595         try:
--> 596             cb(event)
        cb = <bound method EventedList._reemit_child_event of [<Image layer 'astronaut' at 0x7fdabbd3da90>, <Image layer 'photographer' at 0x7fdacc74e790>, <Image layer 'coins' at 0x7fdabc398ed0>, <Image layer 'moon' at 0x7fdabc50b9d0>, <Image layer 'blobs copy' at 0x7fdabf9dec50>, <Image layer 'blobs' at 0x7fdabf9dec90>]>
        event = <Event blocked=False handled=False native=None source=None sources=[] type='translate'>
    597         except Exception:
    598             _handle_exception(

~/GitHub/napari/napari/utils/events/containers/_evented_list.py in _reemit_child_event(self=[<Image layer 'astronaut' at 0x7fdabbd3da90>, <I...9dec50>, <Image layer 'blobs' at 0x7fdabf9dec90>], event=<Event blocked=False handled=False native=None source=None sources=[] type='translate'>)
    183         """An item in the list emitted an event.  Re-emit with index"""
    184         if not hasattr(event, 'index'):
--> 185             setattr(event, 'index', self.index(event.source))
        global setattr = undefined
        event = <Event blocked=False handled=False native=None source=None sources=[] type='translate'>
        self.index = <bound method TypedMutableSequence.index of [<Image layer 'astronaut' at 0x7fdabbd3da90>, <Image layer 'photographer' at 0x7fdacc74e790>, <Image layer 'coins' at 0x7fdabc398ed0>, <Image layer 'moon' at 0x7fdabc50b9d0>, <Image layer 'blobs copy' at 0x7fdabf9dec50>, <Image layer 'blobs' at 0x7fdabf9dec90>]>
        event.source = None
    186         # reemit with this object's EventEmitter of the same type if present
    187         # otherwise just emit with the EmitterGroup itself

~/GitHub/napari/napari/utils/events/containers/_typed.py in index(self=[<Image layer 'astronaut' at 0x7fdabbd3da90>, <I...9dec50>, <Image layer 'blobs' at 0x7fdabf9dec90>], value=<Image layer 'blobs'>, start=0, stop=None)
    240                 "{value!r} is not in list",
    241                 deferred=True,
--> 242                 value=value,
        value = <Image layer 'blobs' at 0x7fdabc6d96d0>
    243             )
    244         )

ValueError: <Image layer 'blobs' at 0x7fdabc6d96d0> is not in list
```

**Comment by tlambert03** [comment]
thanks!

**Comment by tlambert03** [comment]
added a fix.  that's a little bit concerning, so my fix may be more of a patch... but I don't think it's related to this PR, but rather a rarely encountered issue when rapidly popping/adding layers?  not immediately sure.  but fixed here

**Review by sofroniewn — APPROVED** [review]
Fix works for me, I just love this so much! Approving now, but will wait for others to review before merge. Great work @tlambert03. Also once merged I can think we can start thinking about other actions we'd like to add - I already have some ideas!!

**Comment by goanpeca** [comment]
Did a very quick run on it and in general I like the new approach :) 

One thing to consider (for another PR) is how can plugin extend these actions, and how to position them in a given place. On way of doing this could be using something like `insert_before_action(action_unique_name)`.

Having to manually add separators feels a bit off, and something we could eventually consider is having groups/sections and a more programatic way of adding a layer action, instead of passing a dictionary, something like:

```python
action_context_menu = QtActionContextMenu()
action_context_menu.add_action("napari:name", description=trans._("Something"), group="merge", ...)
action_context_menu.add_action("napari:some-other-name", description=trans._("Something else"), group="edit", ...)
```

The order of group creation (for the first time) would define the order of the groups within the menu, and this opens up the possibility of extending the menu with plugins.

**Comment by tlambert03** [comment]
> Having to manually add separators feels a bit off

totally agree...  I like all your suggestions!  

I like `insert_before_action` ... can do that immediately.  I also like the general concept of groups, but not immediately sure how they should acquire their keys.  For this PR, perhaps instead of manually adding separators, if `QtActionContextMenu` is provided with a list of dicts, each dict is considered to be a group?  (with group key "0", "1", etc), or perhaps a tuple of `(group_key, actions_dict)`?

**Comment by goanpeca** [comment]
I think all my suggestions should go to a separate PR, but I wanted to bring that on just in case something needed to be thought a bit harder. So up to you @tlambert03 if you want to add some groundwork for things.

File: napari/layers/base/base.py
Comment (jni) [reply napari/layers/base/base.py:1142]: Ooof. This is really really widely used and also quite nice...

But yeah, `layer = split_channels(viewer.add_image(...), channel_axis=2, colormaps=[...])` is pretty nice and explicit! And honestly not that much more onerous. I'll bring it up at the eurasia meeting to get a temperature read. I think this is probably one of our most used APIs so maybe a longer deprecation like `gui_qt` is in order.

Comment (jni) [reply napari/layers/base/base.py:1142]: fwiw we discussed this at the Eurasia meeting and people were supportive.

Comment (sofroniewn) [reply napari/layers/base/base.py:1142]: Let's save this for a followup PR

**Comment by sofroniewn** [comment]
Let's get this in now, and then we can keep iterating - maybe we turn some of these ideas like `insert_before_action` and dropping `channel_axis` into issues. Will be so great to have this in. I will merge now. Thanks all!!

**Comment by jni** [comment]
Wow! I can't believe this is in! Awesome! :tada: :clap:

**Pull request merged**
